doc/rm/c_cpp_coding_style.md

```diff
@@ -257,6 +257,36 @@ The code using an X Macro must:
     It is usually clearer if you split the definition over multiple lines.
 *   Immediately after the definition, include the `.def` file for the X Macro.
 
+### Undefined Behavior
+
+***All uses of Undefined Behavior (unless stated otherwise) are forbidden without explicit approval of toolchain experts.***
+
+As in any high-quality C/C++ project, Undefined Behavior (as defined by C++14 or C11 as the case may be) must be treated as a security risk.
+To have reasonable hope of being able to upgrade compilers without nasty surprises, we code strictly against the standards.
+
+Although not comprehensive, the following popular idioms are actually Undefined Behavior as defined by the standards.
+This list is provided to illustrate the subtle nature of UB, and how it runs counter to the semantics of C many programmers may have acquired over time.
+
+*   Casting an arbitrary `void *` to `uint32_t *` and dereferencing it.
+    This is the most common accidental violation of the Strict Aliasing Rule.
+*   Creating unaligned pointers.
+    C and C++ consider merely creating such pointers, not just dereferencing them, to be UB.
+*   Comparing (by equality or order) non-null pointers of distinct provenance (e.g. pointers into different stack variables).
+*   Using vanilla `memcpy()` on `volatile` memory; any non-`volatile` access to `volatile` memory is UB.
```

Comment:
Do the current compilers complain about this? If it is necessary to insert a cast to get rid of the warnings from the discarded qualifier then maybe we should talk about casts that discard qualifiers instead?

Reply:
That's a really good question. I got wind of a recent (stalled) patch that would actually make LLVM lower memcpy() of volatile data lower to `@llvm.memcpy.*` with the volatile bit set. @asb thoughts?